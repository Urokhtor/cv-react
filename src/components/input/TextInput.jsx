import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './TextInput.scss';

export default class TextInput extends Component {

    constructor(props) {
        super(props);

        this.state = {
            value: props.value
        };
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        if (prevState.value !== nextProps.value) {
            return {
                value: nextProps.value
            };
        }

        return null;
    }

    onChange(event) {
        this.setState({
            value: event.target.value
        });
    }

    onBlur() {
        this.props.onChange(this.state.value);
    }

    render () {
        return (
            <div className={"text-input flex-vertical"}>
                <input
                    type="text"
                    style={{width: this.props.width}}
                    onChange={this.onChange.bind(this)}
                    onBlur={this.onBlur.bind(this)}
                    value={this.state.value ? this.state.value : ''}
                    required/>
                <label>{this.props.label}</label>
            </div>
        );
    }
}

TextInput.propTypes = {
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string,
    label: PropTypes.string,
    width: PropTypes.number
};

TextInput.defaultProps = {
    width: 300
};