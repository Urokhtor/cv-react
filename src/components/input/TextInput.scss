@import '../../styles/index';

$label-offset: 16px;

.text-input {
  position: relative;

  input {
    background: transparent;

    transition: 0.2s border ease-in-out;

    outline: none;
    border: 0;
    border-bottom: 2px solid $colour-secondary;
    padding-top: $label-offset;
  }

  input:focus {
    border-bottom-color: $colour-accent;
  }

  label {
    position: absolute;
    color: #757575;
    top: $label-offset;
    pointer-events: none;
    transition: all 0.2s ease-in-out;
  }

  input:focus + label, input:valid + label {
    position: absolute;
    top: 0;
    font-size: 50%;
    color: $colour-accent;
    transition: all 0.2s ease-in-out;
  }
}