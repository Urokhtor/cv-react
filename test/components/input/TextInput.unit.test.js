import React from 'react';
import TextInput from "../../../src/components/input/TextInput";

test('TextInput should match snapshot', (done) => {
    const mockOnChange = jest.fn();
    let value = 'some-text-value';
    const wrapper = shallow(<TextInput
            onChange={mockOnChange}
            label={'some-text-label'}
            value={value} />);

    expect(wrapper).toMatchSnapshot();

    done();
});

test('TextInput should return current text value on blur', (done) => {
    const mockOnChange = jest.fn();
    const value = 'some-text-value';
    const wrapper = shallow(<TextInput
        onChange={mockOnChange}
        label={'some-text-label'}
        value={value}/>);

    const input = wrapper.find('input');
    input.simulate('click');
    input.simulate('change', { target: { value: 'new-text-input' }});
    input.simulate('blur');

    expect(mockOnChange).toHaveBeenCalledWith('new-text-input');

    done();
});