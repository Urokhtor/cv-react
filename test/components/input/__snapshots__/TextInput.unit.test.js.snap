// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextInput
    label="some-text-label"
    onChange={[MockFunction]}
    value="some-text-value"
    width={300}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          style={
            Object {
              "width": 300,
            }
          }
          type="text"
          value="some-text-value"
        />,
        <label>
          some-text-label
        </label>,
      ],
      "className": "text-input flex-vertical",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onBlur": [Function],
          "onChange": [Function],
          "required": true,
          "style": Object {
            "width": 300,
          },
          "type": "text",
          "value": "some-text-value",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "some-text-label",
        },
        "ref": null,
        "rendered": "some-text-label",
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            style={
              Object {
                "width": 300,
              }
            }
            type="text"
            value="some-text-value"
          />,
          <label>
            some-text-label
          </label>,
        ],
        "className": "text-input flex-vertical",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onBlur": [Function],
            "onChange": [Function],
            "required": true,
            "style": Object {
              "width": 300,
            },
            "type": "text",
            "value": "some-text-value",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "some-text-label",
          },
          "ref": null,
          "rendered": "some-text-label",
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
